### DISCUSSION TITLE
MOOSE IGA

### DISCUSSION THREAD

**Discussion by makeclean** [idaholab/moose#21642] [Q&A General] [answered]
I've been doing some 'playing' with the recently merged IGA capabilities, which is great, I've been leveraging Coreform Cubit to generate the IGA relevant exodus files. One of the problems I've been playing with is a cylinder problem, one of the things I've noticed is that it I'm not getting the resolution I would expect from the output data, like its correct, but the output resolution seems to be tied to the underlying hex elements, rather than the highly curved elements originating from the IGA. I guess this is because its not clear how VTK should handle it, or is there a remap stage I should go through to get a better representation in plotting the data?

Are other things supported as in 'regular' mesh, such as contact, I guess ray tracing doesn't work? I've not tried, but I'm sure someone is in the know.

**Comment by GiudGiud** [comment]
@roystgnr

**Comment by roystgnr** [comment]
I'm afraid right now there's two alternatives for IGA output: "VTK 9", or "absolutely anything else".  With the "absolutely anything else" option, we output Rational-Bernstein-Bezier control points as if they were Lagrange nodes, and yet no, no they aren't.  It's still useful for debugging a solution or visualizing the coarse features of a solution but not for any kind of post-processing.

With VTK 9 (or possibly 9.1?  see #21449), the Coreform folks added code for us to output rational elements natively in VTK (their own work on that end too! - https://www.kitware.com/implementation-of-rational-bezier-cells-into-vtk/ ).  I'm not sure how robust this is (I was getting failures from it locally, and we don't have the tests running in our regular CI...) however, and on top of that there's the problem that you need your *viz* program to also support rational elements.  IIRC Paraview is the only visualizer that supports rational curves so far, and you might need still need to put in some work on that side for proper support: https://discourse.paraview.org/t/nonlinear-subdivision-for-rational-bezier-curves-not-working/6522

**Reply by makeclean** [reply]
Thanks for the reply, so I would compile libmesh with vtk 9.0, then that *should* be enough to get an appropriate vtk file out - possibly? Then build my own Paraview with vtk 9.0? Looking at one of the linked issues, I guess that would also allow MOOSE to export to VTK using the lagrangian FE elements for better represenation of 2nd order elements and results?

**Reply by msederberg** [reply]
Do these details help? https://mooseframework.inl.gov/modules/tensor_mechanics/examples/cframe_iga.html

**Comment by GregVernon** [comment, marked as answer]
@makeclean - Do you have a screenshot that shows what you're seeing?  I believe it's a known issue atm that the Exodus export from MOOSE isn't supported for these IGA analyses.  Using the VTK output should allow you to specify the nonlinear subdivision level (see below).

To set the nonlinear subdivision level on a ParaView pipeline object:
1.  In the `Properties` browser make sure the `Advanced Properties` are active
![image](https://user-images.githubusercontent.com/868479/180469766-ebb9c88a-42f6-437f-9af3-86fc41828b79.png)
2. Scroll down to the `Miscellaneous` section where you'll see `Nonlinear Subdivision Level`
![image](https://user-images.githubusercontent.com/868479/180469986-62802a0e-eb86-4c9c-a991-d8dc726e5f20.png)
3. I often find `4` to be sufficient and a good compromise with performance

## Example
Nonlinear subdivision level = 1 (default)
![image](https://user-images.githubusercontent.com/868479/180470324-50a7eae2-5394-43a3-a028-9f7daccc7bfe.png)

Nonlinear subdivision level = 4 (my recommendation)
![image](https://user-images.githubusercontent.com/868479/180470380-146775c7-b64d-49fa-8a26-3bb0cdfc975b.png)

## Rendering higher-order elements with edges
There is an outstanding bug regarding rendering higher-order elements as `Surface with Edges` (see discussion [here](https://discourse.paraview.org/t/higher-order-meshes-render-speckled/6108/13) and [here](https://discourse.paraview.org/t/funny-rendering-of-edges-what-is-the-fix/9989)).  The workaround is to essentially create a duplicate of your data and render one as a `Surface` and the other as a `Wireframe`.  Creating the duplicate can either be done with some "zero-op" filters or by loading your data twice.

* `Surface with Edges` bug
![image](https://user-images.githubusercontent.com/868479/180471004-09f255dd-20db-42cd-9364-79cdfd23f664.png)

* `Surface with Edges` workaround
![image](https://user-images.githubusercontent.com/868479/180471168-2738a083-4ef9-4c5f-97bf-60db9fb2b19e.png)

**Reply by makeclean** [reply]
That's massively helpful thanks @GregVernon the only thing I would add to that documentation @msederberg is regarding the configuration of libmesh needing the vtk part, so making sure to build VTK with MPI support and the MPI group support being TRUE, then building libmesh using those libraries, otherwise - very smooth :)
